File: bst.py
class Node:
    __slots__ = 'element', 'left', 'right'

    def __init__(self, element, left=None, right=None):
        self.element = element
        self.left = left
        self.right = right

class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self,troot,e):
        temp = None
        while troot:
            temp = troot
            if e == troot.element:
                return
            elif e < troot.element:
                troot = troot.left
            elif e > troot.element:
                troot = troot.right
        n = Node(e)
        if self.root:
            if e < temp.element:
                temp.left = n
            else:
                temp.right = n
        else:
            self.root = n


    def search(self,key):
        troot=self.root
        while troot:
            if key==troot.element:
                return True
            elif key<troot.element:
                troot=troot.left
            elif key>troot.element:
                troot=troot.right
        
        return False 

    def delete(self,key):
        p=self.root
        pp=None
        while p and p.element!=None:
            pp=p
            if key<p.element:
                p=p.left
            elif key>p.element:
                p=p.right

        if not p:
            return False

        if p.left and p.right:
            s = p.left
            ps = p
            while s.right:
                ps = s
                s = s.right
            p.element = s.element
            p = s
            pp = ps
        c = None
        if p._left:
            c = p.left
        else:
            c = p.right
        if p == self.root:
            self.root = c
        else:
            if p == pp.left:
                pp.left = c
            else:
                pp.right = c

    def inorder(self,troot):
        # print("vinod")
        if troot:
            self.inorder(troot.left)
            print(troot.element,end=' ')
            self.inorder(troot.right)

B = BinarySearchTree()
B.insert(B.root,50)
B.insert(B.root,30)
B.insert(B.root,80)
B.insert(B.root,70)
B.insert(B.root,40)
B.insert(B.root,60)
B.insert(B.root,90)
B.inorder(B.root)
print()
print(B.search(80))
# B.delete(70)
B.inorder(B.root)

File: bst.go
package main

import "fmt"

type Node struct {
	val   int
	left  *Node
	right *Node
}

func insert(root *Node, val int) *Node {
	if root == nil {
		return &Node{val: val}
	}
	if val < root.val {
		root.left = insert(root.left, val)
	} else {
		root.right = insert(root.right, val)
	}
	return root
}

func search(root *Node, val int) bool {
	if root == nil {
		return false
	}
	if root.val == val {
		return true
	}
	if val < root.val {
		return search(root.left, val)
	}
	return search(root.right, val)
}

func delete(root *Node, val int) *Node {
	if root == nil {
		return root
	}
	if val < root.val {
		root.left = delete(root.left, val)
	} else if val > root.val {
		root.right = delete(root.right, val)
	} else {
		if root.left == nil {
			return root.right
		} else if root.right == nil {
			return root.left
		}
		minNode := root.right
		for minNode.left != nil {
			minNode = minNode.left
		}
		root.val = minNode.val
		root.right = delete(root.right, minNode.val)
	}
	return root
}

func main() {
	root := &Node{val: 5}
	insert(root, 3)
	insert(root, 8)
	insert(root, 1)
	insert(root, 4)
	insert(root, 7)
	insert(root, 10)

	fmt.Println(search(root, 3))  // true
	fmt.Println(search(root, 10)) // true
	fmt.Println(search(root, 11)) // false

	root = delete(root, 8)
	fmt.Println(search(root, 8)) // false
}
